static_frame/core/index_hierarchy.py

```diff
@@ -1794,69 +2521,89 @@ class IndexHierarchyGO(IndexHierarchy):
+        if len(value) != self.depth:
+            raise RuntimeError(
+                f'key length ({len(value)}) does not match hierarchy depth ({self.depth})'
+            )
 
-    def extend(self, other: IndexHierarchy) -> None:
-        '''
-        Extend this IndexHiearchy in-place
-        '''
-        self._levels.extend(other._levels)
+        for depth, label_at_depth in enumerate(value):
+            if label_at_depth not in self._indices[depth]:
+                self._indices[depth].append(label_at_depth)
+
+        if self._pending_extensions is None:
+            self._pending_extensions = []
+
+        self._pending_extensions.append(PendingRow(value))
         self._recache = True
 
-    def __copy__(self: IH) -> IH:
+    # 2/3 index.difference
+    # 1/3 index.extend
+    def extend(self: IHGO,
+            other: IndexHierarchy,
+            ) -> None:
         '''
-        Return a shallow copy of this IndexHierarchy.
+        Extend this IndexHierarchyGO in-place
         '''
-        if self._recache:
-            self._update_array_cache()
+        # We do not check whether nor the key exists, as that is too expensive.
+        # Instead, we delay failure until _recache
+
+        for self_index, other_index in zip(self._indices, other._indices):
+            # This will force a recache -> but that's okay.
+            difference = other_index.difference(self_index)
+            if difference.size:
+                self_index.extend(difference)
+
+        if self._pending_extensions is None:
+            self._pending_extensions = []
+
+        self._pending_extensions.append(other)
+        self._recache = True
 
-        blocks = self._blocks.copy()
-        return self.__class__(
-                levels=self._levels.to_index_level(), # recursive deep copy
-                name=self._name,
-                blocks=blocks,
-                own_blocks=True,
-                )
 
 # update class attr on Index after class initialization
 IndexHierarchy._MUTABLE_CONSTRUCTOR = IndexHierarchyGO
 
 
 class IndexHierarchyAsType:
 
-    __slots__ = ('container', 'key',)
+    __slots__ = (
+            'container',
+            'key',
+            )
+
+    container: IndexHierarchy
+    key: GetItemKeyType
 
-    def __init__(self,
+    def __init__(self: IHAsType,
             container: IndexHierarchy,
             key: GetItemKeyType
             ) -> None:
         self.container = container
         self.key = key
 
-    def __call__(self, dtype: DtypeSpecifier) -> IndexHierarchy:
-
+    def __call__(self: IHAsType,
+            dtype: DtypeSpecifier,
+            ) -> IndexHierarchy:
+        '''
+        Entrypoint to `astype` the container
+        '''
         from static_frame.core.index_datetime import dtype_to_index_cls
         container = self.container
 
         if container._recache:
-            container._update_array_cache()
+            container._update_array_cache() # pragma: no cover
```

Comment:
Why no coverage?

Reply:
See this response: https://github.com/InvestmentSystems/static-frame/pull/437#discussion_r831501537